### PR TITLE
Improve minimap camera controls

### PR DESCRIPTION
**Pull request by Zakru**
* Minimap no longer registers camera input when mouse is pressed down elsewhere and brought over the minimap
* Minimap camera responds to press instead of release
* Camera can now be dragged across the minimap in addition to clicking

I felt that at least with the current system, these new camera controls don't fit the existing event model, so I separated them to handle raw mouse input.

### CODE REVIEW

File: crates/controller/src/hud/minimap/interaction.rs [hunk 1]
```diff
@@ -105,15 +119,62 @@ fn click_handler(
     }
 }
 
-fn move_camera_system(
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:108]: I think that having a separate code for camera movement & rest of minimap interaction is not wise. For example unit sending with right clicks suffers the same issues.

I agree that moving the camera on mouse release even when mouse was pressed outside of the minimap is not ideal. Maybe we can fix this by renaming `MinimapClickEvent` to `MinimapPressEvent` and changing `click_handler()` so that it sends the event on press. That would fix the issue with minimal effort.

File: crates/controller/src/hud/minimap/interaction.rs [hunk 2]
```diff
@@ -105,15 +119,62 @@ fn click_handler(
...
+fn move_camera_drag_system(
+    window_query: Query<&Window, With<PrimaryWindow>>,
+    mut input_events: EventReader<MouseMotion>,
+    hud: HudNodes<With<MinimapNode>>,
+    bounds: Res<MapBounds>,
+    mouse_buttons: Res<Input<MouseButton>>,
+    dragging: Res<DraggingCamera>,
+    mut camera_events: EventWriter<MoveFocusEvent>,
+) {
+    input_events.clear();
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:165]: Why is this here?

File: crates/controller/src/hud/minimap/interaction.rs [hunk 3]
```diff
@@ -105,15 +119,62 @@ fn click_handler(
...
+            dragging.0 = true;
+            let proportional = Vec2::new(relative.x, 1. - relative.y);
+            let event = MoveFocusEvent::new(bounds.rel_to_abs(proportional));
+
+            camera_events.send(event);
+        }
+    }
+}
+
+fn move_camera_drag_system(
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:156]: - I think that dragging implementation is independent on the fix to mouse click / press / release behavior. Therefore, I would separate this changes to a standalone system.
- I would change this system to `dragg_handler()` (similar to `click_handler()`) sending `MinimapDraggEvent` and handle the events in `move_camera_system()`.

File: crates/controller/src/hud/minimap/interaction.rs [hunk 4]
```diff
@@ -105,15 +119,62 @@ fn click_handler(
...
+    window_query: Query<&Window, With<PrimaryWindow>>,
+    mut input_events: EventReader<MouseMotion>,
+    hud: HudNodes<With<MinimapNode>>,
+    bounds: Res<MapBounds>,
+    mouse_buttons: Res<Input<MouseButton>>,
+    dragging: Res<DraggingCamera>,
+    mut camera_events: EventWriter<MoveFocusEvent>,
+) {
+    input_events.clear();
+    if !mouse_buttons.pressed(MouseButton::Left) || !dragging.0 {
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:166]: I think that we do not need `dragging` resource at all. Instead we can simplify this by having a `Local<Vec2>` storing the last mouse position and sending the drag event when the current mouse position is different to the last one while `mouse_buttons.pressed(MouseButton::Left)`.

**Review by Indy2222 — CHANGES_REQUESTED** [review]

File: crates/controller/src/hud/minimap/interaction.rs (hunk 2 above)
Comment (Zakru) [reply crates/controller/src/hud/minimap/interaction.rs:165]: Ah, I was somehow under the misconception that `on_event` requires the system to use the events 😳

File: crates/controller/src/hud/minimap/interaction.rs (hunk 4 above)
Comment (Zakru) [reply crates/controller/src/hud/minimap/interaction.rs:166]: We need to be able to tell whether the click started inside the minimap, so that we don't react to a drag that was started outside of it. In fact, if we generalize the drag events as well, I believe we need to store a list/set of the held buttons to keep track of each one separately. What do you think?

File: crates/controller/src/hud/minimap/interaction.rs (hunk 2 above)
Comment (Indy2222) [reply crates/controller/src/hud/minimap/interaction.rs:165]: No, the old events clear automatically.

File: crates/controller/src/hud/minimap/interaction.rs [hunk 5]
```diff
@@ -63,50 +76,117 @@ impl MinimapClickEvent {
...
+struct DraggingButtons(Vec<MouseButton>);
+
 fn click_handler(
     window_query: Query<&Window, With<PrimaryWindow>>,
     mut input_events: EventReader<MouseButtonInput>,
     hud: HudNodes<With<MinimapNode>>,
     bounds: Res<MapBounds>,
-    mut click_events: EventWriter<MinimapClickEvent>,
+    mut dragging: ResMut<DraggingButtons>,
+    mut click_events: EventWriter<MinimapPressEvent>,
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:116]: I would finish with the renaming:

* `click_handler()` -> `press_handler()`
* `mut click_events` -> `mut press_events`
* `InteractionSet::ClickHandler` -> `InteractionSet::PressHandler`

File: crates/controller/src/hud/minimap/interaction.rs [hunk 6]
```diff
@@ -19,35 +22,45 @@ pub(super) struct InteractionPlugin;
...
+                InputSchedule,
+                (
+                    click_handler
+                        .in_set(InteractionSet::ClickHandler)
+                        .run_if(on_event::<MouseButtonInput>()),
+                    drag_handler
+                        .in_set(InteractionSet::DragHandler)
+                        .after(InteractionSet::ClickHandler)
+                        .run_if(on_event::<MouseMotion>()),
+                    move_camera_system.after(InteractionSet::DragHandler),
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:38]: Even though proper ordering is implied, I would be more explicitly here and not depend on other system dependencies: `.after(InteractionSet::ClickHandler).after(InteractionSet::DragHandler)`.

File: crates/controller/src/hud/minimap/interaction.rs [hunk 7]
```diff
@@ -63,50 +76,117 @@ impl MinimapClickEvent {
...
+    mut click_events: EventWriter<MinimapPressEvent>,
 ) {
     let Some(cursor) = window_query.single().cursor_position() else {
         return;
     };
+
     for event in input_events.iter() {
-        if event.state != ButtonState::Released {
+        if event.state != ButtonState::Pressed {
+            dragging.0.retain(|b| *b != event.button);
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:124]: This must be done before the early return (let-else above), otherwise the button would stay "pressed" indefinitely if released outside of the game window.

Also, please test that it works correctly when:

1. press the button while in the minmap
1. hold the button and switch to a different window (e.g. with Alt-TAB)
1. release the button
1. switch back to the game

File: crates/controller/src/hud/minimap/interaction.rs [hunk 8]
```diff
@@ -63,50 +76,117 @@ impl MinimapClickEvent {
...
+        }
+
+        let event = MoveFocusEvent::new(drag.position());
+        camera_events.send(event);
     }
 }
 
 fn send_units_system(
-    mut click_events: EventReader<MinimapClickEvent>,
+    mut click_events: EventReader<MinimapPressEvent>,
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:189]: `click_events` -> `press_events`

File: crates/controller/src/hud/minimap/interaction.rs [hunk 9]
```diff
@@ -118,7 +198,7 @@ fn send_units_system(
 }
 
 fn delivery_location_system(
-    mut click_events: EventReader<MinimapClickEvent>,
+    mut click_events: EventReader<MinimapPressEvent>,
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:201]: `press_events`

File: crates/controller/src/hud/minimap/interaction.rs [hunk 10]
```diff
@@ -63,50 +76,117 @@ impl MinimapClickEvent {
...
+        for b in &dragging.0 {
+            let event = MinimapDragEvent::new(*b, world);
+            drag_events.send(event);
+        }
+    }
+}
+
 fn move_camera_system(
-    mut click_events: EventReader<MinimapClickEvent>,
+    mut click_events: EventReader<MinimapPressEvent>,
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:165]: `click_events` -> `press_events`

File: crates/controller/src/hud/minimap/interaction.rs [hunk 11]
```diff
@@ -63,50 +76,117 @@ impl MinimapClickEvent {
...
 }
 
+fn drag_handler(
+    window_query: Query<&Window, With<PrimaryWindow>>,
+    hud: HudNodes<With<MinimapNode>>,
+    bounds: Res<MapBounds>,
+    dragging: Res<DraggingButtons>,
+    mut drag_events: EventWriter<MinimapDragEvent>,
+) {
+    if dragging.0.is_empty() {
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:145]: Instead of using `dragging.0` (at multiple places), we can use Bevy's `[derive(Deref)]` https://docs.rs/bevy/latest/bevy/prelude/derive.Deref.html

File: crates/controller/src/hud/minimap/interaction.rs [hunk 12]
```diff
@@ -63,50 +76,117 @@ impl MinimapClickEvent {
...
+
+    let Some(cursor) = window_query.single().cursor_position() else {
+        return;
+    };
+
+    if let Some(relative) = hud.relative_position(cursor) {
+        let proportional = Vec2::new(relative.x, 1. - relative.y);
+        let world = bounds.rel_to_abs(proportional);
+
+        for b in &dragging.0 {
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:157]: `b` -> `button` would be more readable.

**Review by Indy2222 — CHANGES_REQUESTED** [review]
I've left multiple comments, but nothing major.

File: crates/controller/src/hud/minimap/interaction.rs (hunk 7 above)
Comment (Zakru) [reply crates/controller/src/hud/minimap/interaction.rs:124]: At least on Windows, when I click -> alt-tab -> release, I don't seem to be receiving a mouse released event, unless the mouse is over the window.
On the other hand when I simply drag from the window to the outside, I consistently receive a release event, but `cursor_position` is inconsistently sometimes `None` and sometimes `Some` with the correct coordinates outside the window.
Any tricks for this?

Comment (Indy2222) [reply crates/controller/src/hud/minimap/interaction.rs:124]: What about checking current state (i.e. Bevy's `Input<MouseButton>::.pressed(MouseButton::Left)`) & combining that with flag whether the last press event occurred over the minimap (every time the mouse button is click, we would update en enum of where it was, i.e. `LastPressLocation::Minimap`)?

Comment (Zakru) [reply crates/controller/src/hud/minimap/interaction.rs]: `Input<MouseButton>` ends up having the same shortcomings, as it is internally being updated using the same `MouseButtonInput` events. Seems like we almost need global input if we want to handle this case "perfectly".

File: crates/controller/src/hud/minimap/interaction.rs [hunk 13]
```diff
@@ -63,68 +78,137 @@ impl MinimapClickEvent {
...
+    mut press_events: EventWriter<MinimapPressEvent>,
 ) {
-    let Some(cursor) = window_query.single().cursor_position() else {
-        return;
-    };
+    let cursor = window_query.single().cursor_position();
+
     for event in input_events.iter() {
-        if event.state != ButtonState::Released {
+        if event.state != ButtonState::Pressed {
```
Comment (Indy2222) [inline crates/controller/src/hud/minimap/interaction.rs:123]: I would `match event.state` so that it doesn't break when other `ButtonState`s are added in the future.

**Review by Indy2222 — CHANGES_REQUESTED** [review]
One minor note, otherwise LGTM.

File: crates/controller/src/hud/minimap/interaction.rs (hunk 7 above)
Comment (Indy2222) [reply crates/controller/src/hud/minimap/interaction.rs]: I am ok with merging it as is. Can you please open a new issue so that we fix that later?

Comment (Zakru) [reply crates/controller/src/hud/minimap/interaction.rs]: I was thinking the same.

**Comment by Zakru** [comment]
Opened #715 based on release event issues

**Review by Indy2222 — APPROVED** [review]
